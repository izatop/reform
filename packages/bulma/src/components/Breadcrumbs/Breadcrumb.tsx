import * as React from "react";
import {XProps} from "../../interfaces";
import {IsActive} from "../../props";
import {MakeProps} from "../../type";
import {ElementFactory} from "../../utils";

const config = ElementFactory.create({displayName: "breadcrumb"});
export const Breadcrumb = config.factory<MakeProps<IsActive>, XProps<"li">>(({props, children}) => (
    <li {...props}>{children}</li>
));
