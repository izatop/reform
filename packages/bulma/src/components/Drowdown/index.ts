export * from "./Dropdown";
export * from "./DropdownDivider";
export * from "./DropdownElement";
