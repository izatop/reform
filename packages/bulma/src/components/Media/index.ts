export * from "./Media";
export * from "./MediaContent";
export * from "./MediaLeft";
export * from "./MediaRight";
