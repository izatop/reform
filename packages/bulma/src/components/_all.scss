@charset "utf-8";

@import "Breadcrumbs/all";
@import "Card/all";
@import "Drowdown/all";
@import "Level/all";
@import "Media/all";
@import "Menu/all";
@import "Message/all";
@import "Modal/all";
@import "Navbar/all";
@import "Pagination/all";
@import "Panel/all";
@import "Tabs/all";
