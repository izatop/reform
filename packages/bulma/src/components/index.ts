export * from "./Breadcrumbs";
export * from "./Card";
export * from "./Drowdown";
export * from "./Level";
export * from "./Media";
export * from "./Menu";
export * from "./Message";
export * from "./Modal";
export * from "./Navbar";
export * from "./Pagination";
export * from "./Panel";
export * from "./Tabs";
