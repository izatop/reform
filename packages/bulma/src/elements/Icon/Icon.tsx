import * as React from "react";
import {IsColor, IsSize} from "../../props";
import {MakeProps} from "../../type";
import {ElementFactory} from "../../utils";
import {FontAwesome} from "./FontAwesome";

export type IconWeight = "lg" | "2x" | "3x";

export interface IIconOptions {
    name: string;
    weight?: IconWeight | string;
    flip?: "v" | "h" | string;
    rotate?: number;
}

export interface IIcon {
    "icon"?: string | IIconOptions;
}

export interface IIconInput extends IsSize, IsColor {
    right?: true;
    left?: true;
}

const config = ElementFactory.create({
    component: "icon",
    resolvers: {color: (v) => `has-text-${v}`},
});

declare const foo: MakeProps<IIconInput>

export const Icon = config.factory<MakeProps<IIconInput>, IIcon>(({props: {className, ...props}, children}) => (
    <>
        <span className={className}><FontAwesome {...props}/></span>
        {React.Children.count(children) > 0 && <span>{children}</span>}
    </>
));
