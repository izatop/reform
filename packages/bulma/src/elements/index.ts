export * from "./Button";
export * from "./Icon";
export * from "./Figure";
export * from "./Notification";
export * from "./Progress";
export * from "./Table";
export * from "./Tag";
export * from "./Title";
export * from "./Box";
export * from "./Container";
export * from "./Content";
export * from "./Delete";
export * from "./Node";
