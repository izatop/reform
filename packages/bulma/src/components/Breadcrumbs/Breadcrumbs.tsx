import * as React from "react";
import {Icon} from "../../elements/Icon";
import {MakeProps, XProps} from "../../interfaces";
import {Size} from "../../options";
import {IsSize} from "../../props";
import {ConfigFactory} from "../../utils";
import {Breadcrumb} from "./Breadcrumb";

export type BreadcrumbsStyleType = "arrow" | "bullet" | "dot" | "succeeds";
export type BreadcrumbNode = React.ReactChild | {} | string;
export type BreadcrumbPath = [string, BreadcrumbNode, string?] | BreadcrumbNode;

export type BreadcrumbsProps = MakeProps<{
    separator?: BreadcrumbsStyleType;
    centered?: boolean;
} & IsSize<"small" | "medium" | "large">>;

export interface IBreadcrumbsAttr extends XProps<"nav"> {
    paths?: BreadcrumbPath[];
}

const config = ConfigFactory.create({
    component: "breadcrumb",
    resolvers: {
        separator: (v) => `has-${v}-separator`,
    },
});

const renderIcon = (icon?: string | null | React.ReactElement) => {
    if (React.isValidElement(icon)) {
        return icon;
    }

    if (typeof icon === "string") {
        return <Icon icon={icon as string} is-size={Size.Small}/>;
    }

    return null;
};

const renderNode = (path: BreadcrumbPath) => {
    if (Array.isArray(path)) {
        const [uri, children, icon = null] = path;
        return (
            <a href={uri}>
                {renderIcon(icon)}
                <span>{children}</span>
            </a>
        );
    }

    if (React.isValidElement(path)) {
        return path;
    }

    return <a>{path}</a>;
};

export const Breadcrumbs = config.factory<BreadcrumbsProps, IBreadcrumbsAttr>(({props, children}) => {
    const {paths, ...p} = props;
    const breadcrumbs = React.useMemo(() => (paths
            ? paths.map(renderNode)
            : React.Children.toArray(children)
    ), [children, props.paths]);

    const lastChild = breadcrumbs.length - 1;

    return (
        <nav aria-label={"breadcrumbs"} {...p}>
            <ul>
                {breadcrumbs.map((child, key) => (
                    <Breadcrumb active={key === lastChild} key={key}>{child}</Breadcrumb>
                ))}
            </ul>
        </nav>
    );
});
