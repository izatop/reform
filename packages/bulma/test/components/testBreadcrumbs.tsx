import * as React from "react";
import * as renderer from "react-test-renderer";
import {Breadcrumbs} from "../../src/components/Breadcrumbs";
import {Icon} from "../../src/elements/Icon";

test("Breadcrumbs", () => {
    const paths = [["/", "Home", "home"], ["/path", "Path"]];
    const element1 = renderer.create(<Breadcrumbs separator={"dot"} paths={paths}/>);
    expect(element1.toJSON()).toMatchSnapshot();

    const element2 = renderer.create((
        <Breadcrumbs centered>
            <Icon icon={"home"}><a href={"/"}>Home</a></Icon>
            <a href={"/path"}>Path</a>
        </Breadcrumbs>
    ));

    expect(element2.toJSON()).toMatchSnapshot();
});
