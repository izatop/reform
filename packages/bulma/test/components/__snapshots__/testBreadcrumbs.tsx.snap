// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs 1`] = `
<nav
  aria-label="breadcrumbs"
  className="breadcrumb has-dot-separator"
>
  <ul />
</nav>
`;

exports[`Breadcrumbs 2`] = `
<nav
  aria-label="breadcrumbs"
  className="breadcrumb is-centered"
>
  <ul>
    <li
      className=""
    >
      <span
        className="icon"
      >
        <i
          className="fas"
        />
      </span>
      <span>
        <a
          href="/"
        >
          Home
        </a>
      </span>
    </li>
    <li
      className=""
    >
      <a
        href="/path"
      >
        Path
      </a>
    </li>
  </ul>
</nav>
`;
